doc/pdmenu.man
.\" -*- nroff -*-
.TH PDMENU 1 "October 29 1999" "1.2.50" "Commands"
.SH NAME
pdmenu \- simple full screen menu program
.SH SYNOPSIS
.B pdmenu [\fIoptions\fP] [\fImenufile\fP ...]
.SH DESCRIPTION
.I pdmenu 
is a simple menu program that displays a menu from which the 
user can pick programs to run. Submenus are supported.
.PP
When you run 
.I pdmenu
, you will be presented with a menu. Simply use the arrow
keys to move to the program you want to run, and press Enter to run the
program. When the program ends, you will be returned to the menu.
.PP
If you are running 
.I pdmenu
at the linux console, and gpm is running, you can
move the mouse and click to navigate through the menus.
.PP
Some menu items are actually submenus, and will take you to another menu.
Others may cause a text edit window to be displayed, where you can enter
parameters for a command before running it. Still others may run a command,
and display the output in a window.
.SH KEYS
Here are all keys you need to get around in \fIpdmenu\fP:
.IP "[up arrow], 8, -"
Move up one line in the menu.
.IP "[down arrow], 2, +"
Move down one line.
.IP "[page up], [ctrl-u]"
Move up an entire screen.
.IP "[page down], [space], [ctrl-v]"
Move down an entire screen.
.IP "[home]"
Move to the first entry of the menu.
.IP "[end]"
Move to the last entry of the menu.
.IP "q"
Exit the current menu, or exit
.I pdmenu 
if used on the first menu. Does not
work in the text edit windows. (This will not work if there is also a hotkey 
set up for \*(L'q\*(R', the hotkey takes precedence.)
.IP "ESC"
Close the currently active window, and return to the previous window, or
exit 
.I pdmenu
if used on the first menu.
.IP "[ctrl-c]"
Exit 
.I pdmenu
immediately.
.IP "[ctrl-r]"
Force a redraw of the screen.
.IP "[enter]"
Launch the selected menu item. In a text edit window, closes the window.
.IP "[backspace]"
Performs a destructive backspace in a text edit window. 
.IP "hotkeys"
Some letters of a menu item may be highlighted. These are hotkeys; simply
press the highlighted letter to select the next menu item with that hotkey.
.SH OPTIONS
.IP "-h, --help"
Display usage summary and exit.
.IP "-c, --color"
Use color. By default, 
.I pdmenu 
will display in black and white mode. If your 
terminal supports color, use this switch.
.IP "-u, --unpark"
"Unparks" the cursor from the bottom of the screen. When this option is
selected, the cursor moves to be on the line of the menu that is currently
selected. This makes 
.I pdmenu
more useable with speech synthesis systems that
need to know what line is the current line on the screen.
.IP "-mmenuid, --menu=menuid"
Instead of displaying the first menu from the menufile, select the menu
with the id "menuid" and display it.
.IP "-q, --quit"
By default, at the opening menu, 'q' will exit \fIpdmenu\fP.
If the \-q switch is
specified, this will not be the case. This is useful if you want to prevent 
the user from ever exiting \fIpdmenu\fP.
.SP
This also disables control-c and the right mouse button from exiting
\fIpdmenu\fP.
.IP "-r, --retro"
This makes
.I pdmenu
use an old style for displaying menus. Menus in the background don't change
color, or lose their hotkeys. Note that this will also be a little bit
faster than the default on slow terminals and the like.
.IP "-l, --lowbit"
By default, 
.I pdmenu
will use nice high bit line drawing characters if it
thinks your terminal is capable. Sometimes it gets this wrong and you get
borders that look all messed up. Then you should use this --lowbit switch to
force it to use low bit line draing characters.
.IP "-n, --numeric"
Disables the use of the keys 2 and 8 for moving up and down. This is useful
if you want to use numbered hotkeys.
.IP "menufile ..."
Specify a menu definition file or files to use. If you specify multiple
files, they will all be loaded in together. By default, the first menu in
the first menufile will be displayed when 
.I pdmenu
starts, but this can be overridden by the --menu= option.
If menufile is "-", pdmenu will read standard input as a menu file.
.SH NOTES
To use the mouse to cut and paste as usual, hold down on the shift key when
you use the mouse.
.SH FILES
.I /usr/local/etc/pdmenurc
.RS
Default config file. See
.BR pdmenurc (5)
for details.
.RE
.I ~/.pdmenurc
.RS
If this exists, it overrides /usr/local/etc/pdmenurc. See
.BR pdmenurc (5)
for details.
.SH ENVIRONMENT
.IP COLORTERM
Set this variable to make
.I pdmenu 
to use color by default.
.SH DISTRIBUTION
Redistribution is subject to the GNU public license.
.SH BUGS
See the file BUGS that came with 
.I pdmenu
for the current buglist.
.SH AUTHOR 
Joey Hess, <joey@kitenet.net>.
.SH "SEE ALSO"
.BR pdmenurc (5)


debian/changelog
pdmenu (1.2.66) unstable; urgency=low

  * Use printf in pdmenurc.debian, not echo -e, so it will work with ash.
  * Copyright updates.
  * Added a title command to the pdmenurc after several user requests.
  * Compiles cleanly with -Wall once more, there were some prototype
    changes in slang.
  * Lost the yellow on red text ("pleasing color scheme"?? what was I
    thinking?)
  * Damn and blast. Debian's patching of slang for wide character set
    support has gone and broken my shadow code, and I don't know how to fix
    it. I've reverted to stupid shadows in the Debian binary. :-(

 -- Joey Hess <joeyh@debian.org>  Fri, 11 May 2001 17:08:38 -0400

pdmenu (1.2.65) unstable; urgency=low

  * Patch from Jan Rafaj <Jan.Rafaj@cedric.vabo.cz> to make it
    open the actual tty, not /dev/tty, if a pdmenurc file is read on stdin.
    The issue is that if /dev/tty is used, child processes use it for
    stdin/out, which, while it should work ok, can yeild confusing behavior
    if you're playing around with ttyname().

 -- Joey Hess <joeyh@debian.org>  Tue, 26 Sep 2000 10:14:01 -0700

pdmenu (1.2.64) unstable; urgency=low

  * Killed the debian/compress file, using -X to dh_compress instead.

 -- Joey Hess <joeyh@debian.org>  Sun, 20 Aug 2000 18:28:38 -0700

pdmenu (1.2.63) unstable; urgency=low

  * Applied a patch from Kees Schoenmakers <KeesS@milltronics.com>
    that is supposed to fix a problem where pdmenu spins in the background
    if the xterm it was running in is closed. I cannot reproduce this one
    myself.

 -- Joey Hess <joeyh@debian.org>  Mon,  7 Aug 2000 22:02:29 -0700

pdmenu (1.2.62) unstable; urgency=low

  * Fixed a dead email address.

 -- Joey Hess <joeyh@debian.org>  Mon, 31 Jul 2000 14:45:03 -0700

pdmenu (1.2.61) unstable; urgency=low

  * Conflicts with older versions of menu that don't support onlyrunasroot.
    Closes: #56744
  * This version will depend on recent versions of libgpm1g, fixing a
    reported segfault with older versions. Closes: #56748

 -- Joey Hess <joeyh@debian.org>  Wed,  5 Apr 2000 17:02:00 -0700

pdmenu (1.2.60) frozen unstable; urgency=low

  * Modified spec file so it will work with .tar.gz files that do not
    include the version number. I am currently generating that type of
    tarball.
  * Merged potato branch. (Words archealogists of the future will long ponder).

 -- Joey Hess <joeyh@debian.org>  Mon,  3 Apr 2000 14:30:44 -0700

pdmenu (1.2.59) frozen unstable; urgency=low

  * Rebuilt to remove cvs cruft "man pages". No code changes. Closes: #58030

 -- Joey Hess <joeyh@debian.org>  Sun, 13 Feb 2000 14:55:09 -0800

pdmenu (1.2.58) frozen unstable; urgency=low

  * Ok, the signal blocking added in the last version just breaks other
    things, so I have reverted it entirely. This version is identical
    to 1.2.56 except for the changelog. Closes: #55651
  * This should go in frozen, since 1.2.57 is broken pretty bad, and this is
    just a reversion to an earlier version.

 -- Joey Hess <joeyh@debian.org>  Wed, 19 Jan 2000 16:13:19 -0800

pdmenu (1.2.57) unstable; urgency=low

  * Block signals before running command, so if it receives a signal while
    the command it running, it doesn't wake up and fight with the command for
    the tty.

 -- Joey Hess <joeyh@debian.org>  Thu, 13 Jan 2000 12:49:53 -0800

pdmenu (1.2.56) unstable; urgency=low

  * Ah, I figured out the shadow weirdness issue at last -- fixed.

 -- Joey Hess <joeyh@debian.org>  Fri,  7 Jan 2000 14:05:34 -0800

pdmenu (1.2.55) unstable; urgency=low

  * Corrected off-by-one error in SHADOW.

 -- Joey Hess <joeyh@debian.org>  Fri,  7 Jan 2000 02:59:05 -0800

pdmenu (1.2.54) unstable; urgency=low

  * Modified menu method to not crash if ran as a user -- it just doesn't do
    anything. Closes: #53933

 -- Joey Hess <joeyh@debian.org>  Wed,  5 Jan 2000 14:09:05 -0800

pdmenu (1.2.53) unstable; urgency=low

  * Build deps added.

 -- Joey Hess <joeyh@debian.org>  Sat,  4 Dec 1999 15:09:54 -0800

pdmenu (1.2.52) unstable; urgency=low

  * Reupload, the source in the debian archive is borked.

 -- Joey Hess <joeyh@debian.org>  Thu,  2 Dec 1999 12:42:01 -0800

pdmenu (1.2.51) unstable; urgency=low

  * Built with a newer slang library, Closes: #50689

 -- Joey Hess <joeyh@debian.org>  Mon, 25 Oct 1999 10:44:53 -0700

pdmenu (1.2.50) unstable; urgency=low

  * Fixed a file descriptior leak. Closes: #46885

 -- Joey Hess <joeyh@debian.org>  Thu,  7 Oct 1999 14:00:10 -0700

pdmenu (1.2.49) unstable; urgency=low

  * First actual code change since february. Fixed a problem with pdmenu not
    closing the mouse when it runs programs. gpm_active was not being set.
    Thanks to Kevin Ryde for pinpointing the problem. (Closes: #46627)

 -- Joey Hess <joeyh@debian.org>  Mon,  4 Oct 1999 15:03:28 -0700

pdmenu (1.2.48) unstable; urgency=low

  * Fixed probable error in redhat spec file, uses FHS man pages now.
    I hope Red Hat can handle FHS..

 -- Joey Hess <joeyh@debian.org>  Mon,  6 Sep 1999 16:47:45 -0700

pdmenu (1.2.47) unstable; urgency=low

  * Fully FHS compliant.

 -- Joey Hess <joeyh@debian.org>  Mon,  6 Sep 1999 16:17:07 -0700

pdmenu (1.2.46) unstable; urgency=low

  * Oops, last upload broke my web page, fixed.

 -- Joey Hess <joeyh@debian.org>  Sat,  4 Sep 1999 22:27:28 -0700

pdmenu (1.2.45) unstable; urgency=low

  * Deleted a lot of accidental cruft.

 -- Joey Hess <joeyh@debian.org>  Sat,  4 Sep 1999 22:24:55 -0700

pdmenu (1.2.44) unstable; urgency=low

  * Corrected CHANGES link.
  * Integrated with my local build system to auto-update home page.

 -- Joey Hess <joeyh@debian.org>  Fri, 27 Aug 1999 20:11:26 -0700

pdmenu (1.2.43) unstable; urgency=low

  * Rebuilt with new slang1-dev to fix slang dependancy.
  * Checked into CVS. Set up a debian/fixlinks script to correct
    symlinks.

 -- Joey Hess <joeyh@debian.org>  Fri, 27 Aug 1999 19:59:12 -0700

pdmenu (1.2.42) unstable; urgency=low

  * Removed redundant part of regexp in debian pdmenurc.


src/pdmenu.c
/*
 * Copyright (c) 1995-2003 Joey Hess (joey@kitenet.net)
 * All rights reserved. See COPYING for full copyright information (GPL).
 */

#include "global.h"
#include "rc.h"
#include "screen.h"
#include "menu.h"
#include "window.h"
#include "mouse.h"
#include "actions.h"
#include "error.h"
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "slang.h"
#include <locale.h>
#include <libintl.h>
#define _(String) gettext (String)

/* Display usage info then quit */
void usage() {
  printf (_("\
Usage: pdmenu [options] [menufiles ..]\n\
\tmenufiles                 the rc files to read instead of\n\
\t                          ~/.pdmenurc or %spdmenurc\n\
\t-h        --help          display this help\n\
\t-c        --color         enable color\n\
\t-u        --unpark        cursor moves to current selection\n\
\t-q        --quit          'q' key does not exit program\n\
\t-mmenuid  --menu=menuid   display menu with this menuid on startup\n\
\t-v        --version       show version information\n\
\t-r        --retro         draw menus using old style\n\
\t-l        --lowbit        do not use high bit line drawing characters\n\
\t-n        --numeric       do not use 8 and 2 for moving up and down\n"),ETCDIR);
#if defined (HAVE_GETOPT_LONG)
#else
  printf ("%s\n", _("(Long options are disabled.)"));
#endif
  exit(-1);
}

/* Display version information then quit */
void version() {
  printf (_("Pdmenu %s GPL Copyright (C) 1995-2002 by Joey Hess <joey@kitenet.net>\n"),VER);
  exit(0);
}

/* 
 * Parse params, Figure out the correct rc file, read it.
 * Returns a pointer to the menu we should display first.
 */
Menu_Type *GetConfig (int argc, char **argv) {
#ifdef HAVE_ASPRINTF
  char **s=malloc(sizeof(char *));
#else
  /* Just long enough for 255 chars of $HOME plus filename. */
  char *s=(char *) malloc(258+strlen("pdmenurc"));
#endif
  char *startmenu=NULL;
  Menu_Type *m;
  int menu_opt_flag=0;
  extern int optind;
  extern char *optarg;
  
#ifdef HAVE_GETOPT_LONG
  struct option long_options[] = {
    {"help",0,NULL,'h'},
    {"color",0,NULL,'c'},
    {"quit",0,NULL,'q'},
    {"unpark",0,NULL,'u'},
    {"version",0,NULL,'v'},
    {"menu",1,NULL,'m'},
    {"retro",0,NULL,'r'},
    {"lowbit",0,NULL,'l'},
    {"numeric",0,NULL,'n'},
    {0, 0, 0, 0}
  };
#endif

#if defined (HAVE_GETOPT_LONG) || defined (HAVE_GETOPT)
  int c=0;

  while (c != -1) {
#ifdef HAVE_GETOPT_LONG
    c=getopt_long(argc,argv,"hcquvm:rln",long_options,NULL);
#elif HAVE_GETOPT
    c=getopt(argc,argv,"hcquvm:rln");
#endif
    switch (c) {
    case 'q': /* 'q' does not exit pdmenu. */
      Q_Exits=0;
      break;
    case 'c': /* Use colors. */
      Use_Color=1;
      break;
    case 'h':
      usage(); /* exits program */
    case 'u': /* Unpark cursor. */
      Unpark_Cursor=1;
      break;
    case 'm': /* Display a particular menu on start. */
      menu_opt_flag=1;
      startmenu=malloc(strlen(optarg)+1);
      strcpy(startmenu,optarg);
      break;
    case 'v':
      version(); /* exits program */
    case 'r':
      Retro=1;
      break;
    case 'l':
      Lowbit=1;
      break;
    case 'n':
	/*
	 * Override the default values for up and down, so 2 and 8 
	 * can be used for hotkeys instead. 
	 */
	Numeric=1;
	break;
    }
  }
#else
  optind=1;
#endif /* have one of the getopts */

  if (optind < argc) { /* we are asked to process a rc file */
    while (optind < argc)
      ReadRc(argv[optind++],RC_FILE);
  }
  else { /* fallback rc files */
#ifdef HAVE_ASPRINTF
    asprintf(s,"%s/.%s",getenv("HOME"),"pdmenurc");
    if ((ReadRc(*s,RC_FILE)==0) || (!menus)) {
      asprintf(s,"%s/%s",ETCDIR,"pdmenurc");
      ReadRc(*s,RC_FILE);
#else
    strncpy(s,getenv("HOME"),255);
    strcat(s,"/.");
    strcat(s,"pdmenurc");
    if ((ReadRc(s,RC_FILE)==0) || (!menus)) {
      strcpy(s,ETCDIR);
      strcat(s,"pdmenurc");
      ReadRc(s,RC_FILE);
#endif
    }
  }

  SanityCheckMenus();
  if (!menus) {
    free(s);
    Error(_("Unable to find any pdmenurc files, or all pdmenurc files are empty."));
  }

  if (menu_opt_flag) { /* menu to display was specified on the command line. */
    m=LookupMenu(startmenu);
    if (m) {
      free(s);
      free(startmenu);
      return m;
    }
    /* Couldn't find the menu if we get to here. */
    Error(_("Menu \"%s\" is not defined."),startmenu);
    exit(1); /* just here to shut up gcc -wall */
  }
  else { /* display first menu we read that still exists */
    free(s);
    /*
     * Sigh I wish we displayed the last menu to appear, then I wouln't 
     * need this loop. Backwards-compatability sucks..
     */
    m=menus;
    while (m->last)
      m=m->last;
    return m;
  }
}

int main (int argc, char **argv) {
  Menu_Type *m;
  int ret;

  setlocale(LC_ALL, "");
  bindtextdomain(PACKAGE,LOCALEDIR);
  textdomain(PACKAGE);
  
  menus=NULL;
  CurrentWindow=NULL;

  ScreenTitle=strdup(_("Pdmenu"));
  
  /* If COLORTERM is set, use color by default */
  Use_Color=!(getenv("COLORTERM")==NULL);
  Q_Exits=1;
  Unpark_Cursor=0;

  /* SLang screen objects and their default colors */
  strcpy(ScreenObjNames[1],DESKTOP_NAME);	
  strcpy(FG[0],DESKTOP_FG_DEFAULT);	
  strcpy(BG[0],DESKTOP_BG_DEFAULT);
  strcpy(ScreenObjNames[6],TITLE_NAME);	
  strcpy(FG[5],TITLE_FG_DEFAULT);	
  strcpy(BG[5],TITLE_BG_DEFAULT);
  strcpy(ScreenObjNames[3],BASE_NAME);	
  strcpy(FG[2],BASE_FG_DEFAULT);	
  strcpy(BG[2],BASE_BG_DEFAULT);
  strcpy(ScreenObjNames[4],MENU_NAME);	
  strcpy(FG[3],MENU_FG_DEFAULT);	
  strcpy(BG[3],MENU_BG_DEFAULT);
  strcpy(ScreenObjNames[5],SELBAR_NAME);	
  strcpy(FG[4],SELBAR_FG_DEFAULT);	 
  strcpy(BG[4],SELBAR_BG_DEFAULT);
  strcpy(ScreenObjNames[2],SHADOW_NAME);	
  strcpy(FG[1],SHADOW_FG_DEFAULT);
  strcpy(BG[1],SHADOW_BG_DEFAULT);
  strcpy(ScreenObjNames[7],MENU_HI_NAME);
  strcpy(FG[6],MENU_HI_FG_DEFAULT);	
  strcpy(BG[6],MENU_HI_BG_DEFAULT);
  strcpy(ScreenObjNames[8],SELBAR_HI_NAME);	
  strcpy(FG[7],SELBAR_HI_FG_DEFAULT);	
  strcpy(BG[7],SELBAR_HI_BG_DEFAULT);
  strcpy(ScreenObjNames[9],UNSEL_MENU_NAME);
  strcpy(FG[8],UNSEL_MENU_FG_DEFAULT);
  strcpy(BG[8],UNSEL_MENU_BG_DEFAULT);
  
  /* Parse parameters and load pdmenurc file. */
  m=GetConfig(argc,argv);

  /* color or b&w? Tell slang */
  if (Use_Color) { /* color */
    DESKTOP=1;
    TITLE=6;
    BASE=3;
    MENU=4;
    SELBAR=5;
    SHADOW=1;
    MENU_HI=7;
    SELBAR_HI=8;
    UNSEL_MENU=9;
  }
  else { /* b&w: those numbered greater than 1 are inverse. */
    DESKTOP=0;
    TITLE=2;
    BASE=2;
    MENU=0;
    SELBAR=2;
    SHADOW=0;
    MENU_HI=2;
    SELBAR_HI=0;
    UNSEL_MENU=0;
  }

  Screen_Setcolors();
  SetScreensize();

#ifdef GPM_SUPPORT
  gpm_ok=InitMouse();
#endif

  AddWindow(m);
  Screen_Init();
  DrawAll();

  do {
    ret=DoMenu(m,Pdmenu_Action,Handle_Ctrl_C);
    if (tcgetpgrp(0) == -1) break; /* Detect tty gone away. */
  } while ((ret!=QUIT_EXIT) && !((ret==Q_KEY_EXIT) && (Q_Exits)));

#ifdef GPM_SUPPORT
  EndMouse(); 
#endif
  Screen_Reset();
  return 0;                  
}
